### ISSUE TITLE
Shorttag to show/hide app specific content

### ISSUE THREAD

**Issue by gerhardsletten** [uncatcrea/wp-appkit#211]
Would a shorttags for showing / hiding content in a post/page be good idea? Probably also to target specific platforms like ios or android?

**Comment by blupu** [comment]
We had in mind to add shorttags to embed alternative content for mobile (notably when it uses Flash). Is that what you have in mind?

**Comment by gerhardsletten** [comment]
Hi, I was thinking more that if you reuse wordpress pages both in the website and for the app, it would be nice to adjust the output based on where you consume it.

My use case now, where a parent site where the site owner had added internal links to sub-pages in the content of the page, but in the app you would rather use the "List subpages option". 

You could add a special css-class "hide-in-app" and wrap that content with it, and then add a css rule in the apps theme to not display that. But a more user-friendly what would be to have a set over shortcodes to show / hide different part of the page-content. The code code could be somethink like this:

```
<?php 
add_shortcode( 'hide_app', 'shorttag_hide_app' ); 
function shorttag_hide_app( $atts, $content ) {
  if ( is_app() ) { // is_app() = some kind of function to check it content is generated for the app
    return '';
  }
  return $content;
}

add_shortcode( 'only_app', 'shorttag_only_app' ); 
function shorttag_only_app( $atts, $content ) {
  if ( is_app() ) {
    return $content;
  }
  return '';
}
?>
```

**Comment by blupu** [comment]
Good point. It could be a nice feature indeed. So I believe the shortcodes would only interpreted when data are retrieved through the WP-AppKit API. @mleroi do you see a way to do the _is_app()_?
Regarding the shorcodes themselves, I would probably have something like _hide_from_apps_ and _show_in_apps_only_ to be as explicit as possible.

**Comment by mleroi** [comment]
Yes _is_app()_ is totally doable, we don't have a dedicated function to to that yet but it can totally be integrated to v0.6 for example.

I made a quick test with the shortcodes implemented as above and defining _is_app()_ like so:
```
function is_app() {
	return !empty( WpakWebServiceContext::$current_app_id );
}
```
It works like a charm :)

**Comment by blupu** [comment]
Great news :-) It will integrated in 0.6.

**Comment by blupu** [comment]
We may have to consider the following: _hide_from_apps_ and _show_only_in_apps_ will strip content. In other words, a content fragment tagged with _hide_from_apps_ will never reach the app itself. It will be deleted from content before the data are returned to apps. However, sometimes we need to signal that a content was there and can't be displayed in the app. It will allow to style a placeholder saying _content unvailable_ (for example).

So I propose to have a _notify_ attribute for the _hide_from_apps_ shortcode with a true|false values or yes|no values. The default value will be false|no (meaning the tagged content is simply stripped and that the app doesn't know about it).

**Comment by mleroi** [comment]
Shortcodes created:
* [show_only_in_apps]content[/show_only_in_apps]
* [hide_from_apps notify="yes/no"]content[/hide_from_apps] (*notify* is optional)

I'm wondering if *show_only_in_apps* could not be named *only_in_apps* (when making tests I found *show_only_in_apps* a bit long to write for a shortcode name). But tell me what you think!

Any test/feedback on this is welcome!

**Comment by blupu** [comment]
@mleroi I've tested shortcodes and they work.

However, I think, we should enhance their outputs. `[show_only_in_apps]` and `[hide_from_apps]` should be able to return their content in `<p>` tags. (Not always as we might want to show/hide inline content.)

To match this new behavior, I would output a `<p>` by default for _.wpak-content-not-available_. If we have a mechanism to set output as inline elements, then _.wpak-content-not-available_ can be returned as `<span>`.

Maybe the easiest way is to have a _container_ attribute whose value is the surrounding tag (eg. span, p, div...). If the _container_ attribute is not set, the default value is `<p>` (ie. block).

**What do you think of that?**

**Comment by mleroi** [comment]
Defaulting to `<p>` and having a _container_ attribute seems a good idea indeed.
I'll add that :)

**Comment by mleroi** [comment]
@blupu I've added the _container_ attribute.
`p` by default and can be set to any html tag, or set to "none" to display no wrapper (as it was before).
New usages:
`[show_only_in_apps container="p|span|div|none"]content[/show_only_in_apps]`
`[hide_from_apps notify="yes|no" container="p|span|div|none"]content[/hide_from_apps]`

**Comment by mleroi** [comment]
@blupu if no _container_ attribute then container defaults to `<p>`

**Comment by mleroi** [comment]
Here's the doc section about those 2 shorcodes "show_only_in_apps" and "hide_from_apps": http://uncategorized-creations.com/wp-appkit/doc/app-content/#2721-shortcodes
